rfcs/62-editor-data-model.md

```diff
@@ -0,0 +1,653 @@
+```
+
+The intent is to explore some optimization strategies like string interning for fast comparison. Those optimizations are left out of the scope of this RFC.
+
+### Built-in types
+
+The data model supports a number of built-in types, corresponding to common Rust built-in types.
+
+```rust
+enum SignedIntegralType {
+    Int8, Int16, Int32, Int64, Int128,
+}
+
+enum UnsignedIntegralType {
+    Uint8, Uint16, Uint32, Uint64, Uint128,
+}
+
+enum FloatingType {
+    Float32, Float64,
+}
+
+enum SimpleType {
+    Bool,
+    Int8, Int16, Int32, Int64, Int128,
+    Uint8, Uint16, Uint32, Uint64, Uint128,
+    Float32, Float64,
+}
+
+enum BuiltInType {
+    Bool,
+    Int8, Int16, Int32, Int64, Int128,
+    Uint8, Uint16, Uint32, Uint64, Uint128,
+    Float32, Float64,
+    Name,
+    Enum(UnsignedIntegralType), // C-style enum, not Rust-style
+    Array(AnyType),
+    Dict(AnyType), // key type is Name
+    ObjectRef(Option<Entity>),
+}
+```
+
+_Note_: We explicitly handle all integral and floating-point bit variants to ensure tight value packing and avoid wasting space both in structs (`GameObject`; see below) and collections (arrays, _etc._).
+
+`ObjectRef` is a reference to any object, custom or built-in. This allows referencing components built-in inside Bevy or the Editor (TBD depending on design of that part). The reference needs to be valid (non-null) when the data gets processed for baking, but can temporarily (including while serialized to disk and later reloaded) be left invalid (`None`) for editing flexibility. Nullable references require a `NullableRef` component to be added to mark the reference as valid even if null.
+
+```rust
+#[derive(Component)]
+struct NullableRef;
+```
+
+This allows filtering out nullable references, and collect invalid `ObjectRef` instances easily via a `Query`, for example to emit some warning to the user.
+
+The `AnyType` is defined as either a built-in type or a reference to a custom Game type (see [Game types](#game-types)).
+
+```rust
+enum AnyType {
+    BuiltIn(BuiltInType),
+    Custom(Entity), // with GameType component
+}
+```
+
+### Game world
+
+The data model uses dynamic types, so it can be represented purely with data, without any build-time types. Bevy already has an ideal storage for data: the `World` ECS container. This RFC introduces a new struct to represent the data model, the `GameWorld`:
+
+```rust
+struct GameWorld(pub World);
+```
+
+Since the underlying `World` already offers guarantees about concurrent mutability prevention, via the Rust language guarantees themselves, those guarantees transfer to the `GameWorld` allowing safe concurrent read-only access or exclusive write access to all data of the data model.
+
+**TODO - Explain where the `GameWorld` lives. Idea was as a resource on the Editor's `World`, but that means regular Editor systems cannot directly write queries into the `GameWorld` ergonomically. See how extract does it for the render world though.**

```

Comment:
What "deal" is there to have with editor types? I'm not sure I'm following.

The `Editor` is a Bevy app so as all Bevy apps have a `MainWorld` _etc._ nothing specific to that RFC. The types and instances of the Game that the Editor edits need to be objects instantiated in the Editor program. Now, we could use a `Vec` or `HashMap` to store them, but we have a largely more powerful container that Bevy gives us: `World`. So we (ab)use `World` to store into what I called a `GameWorld` all that data and efficiently work with it. And because the Editor is a Bevy app, a natural place to store some data in your program is as a `Resource` on your `MainWorld`.

But there no plan for the Editor to manipulate dynamic representations of its own types; I really don't see what we'd do with that.